File: problem2.js

// testing could be random inputs within a range
let string = ")()"

function longestPattern(string) {
  if (string.length === 0) {
    return 0
  }

  let max = 0
  let stack = [-1]

  for (let index = 0; index < string.length; index++) {
    if (string[index] === '(') {
      stack.push(index)
      // for fun: console.log(stack)
    } else {
      stack.pop()

      if (stack.length === 0) {
        stack.push(index)
      } else {
        max = Math.max(max, index - stack[stack.length - 1])
      }

    }
  }
  return max
}

longestPattern(string)

File: problem2.py
class Answer:
    def longestParentheses(self, string: str) -> int:
        zo = 0
        stack = [-1]
        for i in range(len(string)):
            if string[index] == '(':
                stack.append(index)
            else:
                stack.pop()
                if not stack: 
                    stack.append(index)
                else:
                    zo = max(zo, index - stack[-1])
        return zo
